v2/devdoc/log_sink_etw_requirements.md

```diff
@@ -0,0 +1,179 @@
+
+`log_sink_etw` implements a log sink interface that logs the events as ETW events.
+
+## References
+
+[Event Tracing For Windows](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/event-tracing-for-windows--etw-)
+
+## Exposed API
+
+```
+    extern const LOG_SINK_IF log_sink_etw;
+```
+
+### log_sink_etw.log_sink_log
+
+The signature of `log_sink_etw.log_sink_log` is:
+
+```c
+typedef void (*LOG_SINK_LOG_FUNC)(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line, const char* message_format, ...);
+```
+
+`log_sink_etw.log_sink_log` implements logging to an ETW provider.
+
+If `message_format` is `NULL`, `log_sink_etw.log_sink_log` shall fail and return.
+
+`log_sink_etw_log` shall register the ETW TraceLogging provider by calling `TraceLoggingRegister` if it was not already done.
+
+`log_sink_etw_log` shall use as provider GUID `DAD29F36-0A48-4DEF-9D50-8EF9036B92B4`.
+
+Note this can (and should) be improved to be configurable later.
+
+`log_sink_etw_log` shall emit a self described event that shall have the name of the event as follows:
+
+  - If `log_level` is `LOG_LEVEL_CRITICAL` the event name shall be `LogCritical`.
+
+  - If `log_level` is `LOG_LEVEL_ERROR` the event name shall be `LogError`.
+
+  - If `log_level` is `LOG_LEVEL_WARNING` the event name shall be `LogWarning`.
+
+  - If `log_level` is `LOG_LEVEL_INFO` the event name shall be `LogInfo`.
+
+  - If `log_level` is `LOG_LEVEL_VERBOSE` the event name shall be `LogVerbose`.
+
+  - Otherwise the event name shall be `Unknown`.
+
+- Logging level:
+
+  - If `log_level` is `LOG_LEVEL_CRITICAL` the ETW logging level shall be `TRACE_LEVEL_CRITICAL`.
+
+  - If `log_level` is `LOG_LEVEL_ERROR` the ETW logging level shall be `TRACE_LEVEL_ERROR`.
+
+  - If `log_level` is `LOG_LEVEL_WARNING` the ETW logging level shall be `TRACE_LEVEL_WARNING`.
+
+  - If `log_level` is `LOG_LEVEL_INFO` the ETW logging level shall be `TRACE_LEVEL_INFO`.
+
+  - If `log_level` is `LOG_LEVEL_VERBOSE` the ETW logging level shall be `TRACE_LEVEL_VERBOSE`.
+
+  - Otherwise the ETW logging level shall be `TRACE_LEVEL_NONE`.
+
+- If `log_context` is `NULL` only the fields `content`, `file`, `func` and `line` shall be added to the ETW event.
+
+- If `log_context` is not `NULL`:
+
+`log_sink_etw.log_sink_log` shall call `log_context_get_property_value_pair_count` to obtain the count of properties that are to be added to the ETW event.
+
+`log_sink_etw.log_sink_log` shall call `log_context_get_property_value_pairs` to obtain the properties that are to be added to the ETW event.
+
+`log_sink_etw.log_sink_log` shall limit the number of properties that are emitted in the ETW event to 64.
+
+`log_sink_etw.log_sink_log` shall compute the metadata size for the self described event metadata as follows:
+
+- Length of the event name (excluding zero terminator) + 1.
+
+- Length of the `content` field name + 1.
+
+- Length of the `file` field name + 1.
```

Comment:
>Length [](http://example.com/codeflow?start=2&length=6)

Wondering ... since `__FILE__` is a constant... can this length be predetermined for every file? And then no `strlen`?

Reply:
Yes, this will be fixed, no strlen, added to the spec.